### ISSUE TITLE
Create table with foreign key on RocksDB should report error

### ISSUE THREAD

**Issue by tplavcic** [facebook/mysql-5.6#205]
Something like this reports OK but I think it should report as error or at least give warning:
With: --no-defaults --default-tmp-storage-engine=MyISAM --rocksdb --skip-innodb --default-storage-engine=RocksDB

```
mysql> create table test (a int not null, b int not null, foreign key (b) references test2(b));                                                                                                                                  
Query OK, 0 rows affected (0.12 sec)
```

It just creates a table without a foreign key so a user is not aware that the table is not what he wanted to create:

```
mysql> show create table test;
+-------+-------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                  |
+-------+-------------------------------------------------------------------------------------------------------------------------------+
| test  | CREATE TABLE `test` (
  `a` int(11) NOT NULL,
  `b` int(11) NOT NULL,
  KEY `b` (`b`)
) ENGINE=ROCKSDB DEFAULT CHARSET=latin1 |
+-------+-------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

If I understand correctly that foreign keys are not supported then the error or at least warning would be good.

**Comment by mdcallag** [comment]
Wow, MyRocks continues the MySQL tradition of "parsed but ignored"
http://mysqlha.blogspot.com/2009/01/parsed-but-ignored.html

**Comment by yoshinorim** [comment]
The described behavior applies to MyISAM and MEMORY too. I think MySQL layer lacks checking HTON_SUPPORTS_FOREIGN_KEYS at DDL.

**Comment by jkedgar** [comment]
I've looked into a bit.  It looks like MySQL _intentionally_ ignores the foreign key constraint for the MyISAM and MEMORY storage engines.  The key is still created, but the relationship with the other table is just ignored.  Probably the easiest way to get a warning would be to change check_fk_parent_table_access() in sql_parse.cc, but that would then change the behavior in MyISAM and MEMORY, which I don't think we want.  As it is, once the table gets into the handler, all the foreign key constraint stuff has been stripped out.

Instead I wonder if we want to _turn on_ HTON_SUPPORTS_FOREIGN_KEYS which then means MySQL would validate that the referenced table exists as well as all the other checking needed.  Then we would get the foreign key constraint information in the MyRocks handler and we could then generate an error (until such a time as we implement foreign keys).

Thoughts?

**Comment by mdcallag** [comment]
Jay has discovered "parsed but ignored"
http://mysqlha.blogspot.com/2009/01/parsed-but-ignored.html

On Wed, Apr 13, 2016 at 11:57 AM, Jay Edgar notifications@github.com
wrote:

> I've looked into a bit. It looks like MySQL _intentionally_ ignores the
> foreign key constraint for the MyISAM and MEMORY storage engines. The key
> is still created, but the relationship with the other table is just
> ignored. Probably the easiest way to get a warning would be to change
> check_fk_parent_table_access() in sql_parse.cc, but that would then change
> the behavior in MyISAM and MEMORY, which I don't think we want. As it is,
> once the table gets into the handler, all the foreign key constraint stuff
> has been stripped out.
> 
> Instead I wonder if we want to _turn on_ HTON_SUPPORTS_FOREIGN_KEYS which
> then means MySQL would validate that the referenced table exists as well as
> all the other checking needed. Then we would get the foreign key constraint
> information in the MyRocks handler and we could then generate an error
> (until such a time as we implement foreign keys).
> 
> Thoughts?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/facebook/mysql-5.6/issues/205#issuecomment-209595241

## 

Mark Callaghan
mdcallag@gmail.com

**Comment by jkedgar** [comment]
Turns out I did not need to set HTON_SUPPORTS_FOREIGN_KEYS.  It appears that MySQL does not pass any foreign key information to the storage engine.  InnoDB handles it by getting access to the query statement and parsing it looking for the FOREIGN KEY declarations.  I've added something similar to generate an error if FOREIGN KEY declarations are included in CREATE TABLE and ALTER TABLE statements.

https://reviews.facebook.net/D56703
